### PR TITLE
Created a meaningful error message and error handling when files are …

### PR DESCRIPTION
**Pull request by Skarlso**
…inaccessible for an MD5 has computation.

### CODE REVIEW

**Comment by Skarlso** [comment]
Hi, here is the linked google group chat about this pull-request => https://groups.google.com/forum/#!topic/go-cd-dev/N35VPPMrF8o

Cheers,
Gergely.

**Comment by ketan** [comment]
Hi -

Have you signed the contributor license agreement? https://github.com/gocd/gocd/blob/master/CONTRIBUTING.md

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java [hunk 1]
```diff
@@ -44,6 +44,7 @@ public PluginsZip(SystemEnvironment systemEnvironment, ZipUtil zipUtil) {
         bundledPlugins = new File(systemEnvironment.get(SystemEnvironment.PLUGIN_GO_PROVIDED_PATH));
         externalPlugins = new File(systemEnvironment.get(SystemEnvironment.PLUGIN_EXTERNAL_PROVIDED_PATH));
         this.zipUtil = zipUtil;
+        checkFilesAccessibility(bundledPlugins, externalPlugins);
```
Comment (ketan) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: We prefer the constructors to not blow up with unexpected errors. Would you mind moving this check into the `create()` method?

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java [hunk 2]
```diff
@@ -73,7 +74,21 @@ private String computeMd5DigestOfPlugins() {
...
+            throw new RuntimeException(String.format("Could not compute md5 of plugins." +
+                    " Exception occurred: %s", e.getStackTrace()));
         }
     }
+
+    private void checkFilesAccessibility(File bundledPlugins, File externalPlugins) {
+        try {
+            if (!bundledPlugins.canRead() || !externalPlugins.canRead()) {
+                LOG.error("Could not read plugins." +
+                        "Please check access rights on files in external and bundled folder.");
```
Comment (ketan) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: It is ok for lines to be longer than what you have. Feel free to declare this string a single line -

```
                LOG.error("Could not read plugins. Please check access rights on files in external and bundled folder.");
```

A nice improvement would be to print the locations of those folders :)

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java [hunk 3]
```diff
@@ -73,7 +74,21 @@ private String computeMd5DigestOfPlugins() {
...
+                throw new FileAccessRightsCheckException("Could not read plugins. " +
+                        "Please check access rights in external and bundled plugin folder.");
+            }
+        } catch (SecurityException se) {
```
Comment (ketan) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: Are you trying to catch this from `File#canRead()`?

We aren't catching `SecurityException` anywhere in the app, I feel it's OK to drop this catch block.

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java (hunk 1 above)
Comment (Skarlso) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: Sure, done.

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java (hunk 2 above)
Comment (Skarlso) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: Sure. :)

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java (hunk 3 above)
Comment (Skarlso) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: Yes, as a precaution. Okay, I'll drop the catch block.

**Comment by Skarlso** [comment]
Yes. :) Signed!

**Comment by ketan** [comment]
Also, when you're done, could you please squash all commits into a single commit, and let us know when it is ready, so someone can look into it.

Thanks!

**Comment by Skarlso** [comment]
Will do!

You are welcome! I'm happy to help out when I can. :)

G.

**Comment by Skarlso** [comment]
All done. 

Cheers,
Gergely.

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java [hunk 4]
```diff
@@ -73,7 +74,20 @@ private String computeMd5DigestOfPlugins() {
...
+
+    private void checkFilesAccessibility(File bundledPlugins, File externalPlugins) {
+        boolean bundled = bundledPlugins.canRead();
+        boolean external = externalPlugins.canRead();
+        if (!bundled || !external) {
+            String folder = bundled ? externalPlugins.getAbsolutePath() : bundledPlugins.getAbsolutePath();
+            LOG.error(String.format("Could not read plugins." +
+                    "Please check access rights on files in folder: %s.", folder));
+            throw new FileAccessRightsCheckException(String.format("Could not read plugins. " +
+                    "Please check access rights in folder: %s", folder));
```
Comment (zabil) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: Here too ```String.format()``` the string in one line?

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java [hunk 5]
```diff
@@ -73,7 +74,20 @@ private String computeMd5DigestOfPlugins() {
             String digestOfPlugins = digestOfBundledFolder + digestOfExternalFolder;
             return DigestUtils.md5Hex(digestOfPlugins);
         } catch (Exception e) {
-            throw new RuntimeException(String.format("Could not compute md5 of plugins. Exception occurred: %s", e.getStackTrace()));
+            throw new RuntimeException(String.format("Could not compute md5 of plugins." +
+                    " Exception occurred: %s", e.getStackTrace()));
```
Comment (zabil) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: Can we have ```String.format()``` the string in one line?

Comment (Skarlso) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: Yepp, sure.

File: plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java (hunk 4 above)
Comment (Skarlso) [inline plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/commons/PluginsZip.java]: Certainly.

**Comment by zabil** [comment]
Thanks!

**Pull request merged**